### PR TITLE
[WIP] Payment channel example.

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

File: docs/solidity-by-example.rst [hunk 1]
```diff
@@ -620,4 +620,52 @@ Safe Remote Purchase
...
+    contract Channel {
+        address[2] users;
+        uint deposit;
+        uint timeout;
+        uint closeTimestamp;
+        uint closeSequenceNumber;
+        uint closeBalanceUser0;
+        enum State { Opened, Accepted, Closing, Closed }
+        State state;
+        function Channel(address _other, uint _timeout) payable {
```
Comment (axic) [inline docs/solidity-by-example.rst]: I think some whitespaces between functions, state variables and the pragma cannot hurt.

File: docs/solidity-by-example.rst [hunk 2]
```diff
@@ -620,4 +620,52 @@ Safe Remote Purchase
...
+            users[0] = msg.sender;
+            users[1] = _other;
+            deposit = msg.value;
+            require(deposit < 2**250);
+            require(_timeout < 2 days);
+            timeout = _timeout;
+        }
+        function accept() payable {
+            require(msg.sender == users[1]);
```
Comment (axic) [inline docs/solidity-by-example.rst]: Since `user[0]` and `user[1]` have a very specific role, I think it would make sense naming them appropriately.

Comment (chriseth) [reply docs/solidity-by-example.rst]: Initiator and acceptor?

**Comment by chriseth** [comment]
This is work in progress and requires much more documentation. You could already check whether the contract is a valid payment channel, though.

File: docs/solidity-by-example.rst (hunk 2 above)
Comment (axic) [reply docs/solidity-by-example.rst]: That sounds fine.

File: docs/solidity-by-example.rst [hunk 3]
```diff
@@ -620,4 +620,52 @@ Safe Remote Purchase
...
+            require(msg.value == deposit);
+            require(state == State.Opened);
+            state = State.Accepted;
+        }
+        function requestClose(uint _balanceUser0, uint _sequenceNumber, uint8 v, bytes32 r, bytes32 s) {
+            require(state == State.Accepted || state == State.Closing);
```
Comment (axic) [inline docs/solidity-by-example.rst]: Why allow closing multiple times? Each time the timestamp/balance/sequence is updated.

**Comment by axic** [comment]
I suggest not to merge this before #2553.

**Comment by axic** [comment]
Rebased to trigger the tests.

File: docs/solidity-by-example.rst [hunk 4]
```diff
@@ -620,4 +620,52 @@ Safe Remote Purchase
...
+            require(_timeout < 2 days);
+            timeout = _timeout;
+        }
+        function accept() payable {
+            require(msg.sender == users[1]);
+            require(msg.value == deposit);
+            require(state == State.Opened);
+            state = State.Accepted;
+        }
+        function requestClose(uint _balanceUser0, uint _sequenceNumber, uint8 v, bytes32 r, bytes32 s) {
```
Comment (axic) [inline docs/solidity-by-example.rst]: Inconsistency: v/r/s is not prefixed.

File: docs/solidity-by-example.rst (hunk 3 above)
Comment (axic) [reply docs/solidity-by-example.rst]: @chriseth this is still an open question

**Comment by axic** [comment]
@chriseth updating this with the comments. Can you review? What is left to merge this?

**Comment by axic** [comment]
@chriseth rebased. What is blocking this?

File: docs/solidity-by-example.rst [hunk 5]
```diff
@@ -619,4 +619,63 @@ Safe Remote Purchase
...
+        function counterpartyAddress(uint _sequenceNumber) internal returns (address) {
+            return (_sequenceNumber % 2 == 1) ? acceptor : initiator;
+        }
+
+        function requestClose(uint _balanceInitiator, uint _sequenceNumber, uint8 _v, bytes32 _r, bytes32 _s) {
+            require(state == State.Accepted || state == State.Closing);
+            require(_sequenceNumber > closeSequenceNumber);
+            require(_balanceInitiator <= 2 * deposit);
+            require(closeTimestamp == 0 || now <= closeTimestamp + timeout);
+            require(ecrecover(sha3(_balanceInitiator, _sequenceNumber), _v, _r, _s) == counterpartyAddress(_sequenceNumber));
```
Comment (axic) [inline docs/solidity-by-example.rst:663]: Should use `keccak256`.

File: docs/solidity-by-example.rst [hunk 6]
```diff
@@ -619,4 +619,63 @@ Safe Remote Purchase
...
+        }
+
+        function accept() payable {
+            require(msg.sender == acceptor);
+            require(msg.value == deposit);
+            require(state == State.Opened);
+            state = State.Accepted;
+        }
+
+        function counterpartyAddress(uint _sequenceNumber) internal returns (address) {
```
Comment (axic) [inline docs/solidity-by-example.rst]: Should use `pure`.

**Comment by axic** [comment]
Compilation failures:
```
Warning: This is a pre-release compiler version, please do not use it in production.
/tmp/tmp.aEq6oBGgwl/test_0a4d82f0af2c945e6a612af13a590776a772d071cdf8cbc0cce41dba326a8ffb.sol:14:9: Warning: No visibility specified. Defaulting to "public".
        function PaymentChannel(address _other, uint _timeout) payable {
        ^
Spanning multiple lines.
/tmp/tmp.aEq6oBGgwl/test_0a4d82f0af2c945e6a612af13a590776a772d071cdf8cbc0cce41dba326a8ffb.sol:23:9: Warning: No visibility specified. Defaulting to "public".
        function accept() payable {
        ^
Spanning multiple lines.
/tmp/tmp.aEq6oBGgwl/test_0a4d82f0af2c945e6a612af13a590776a772d071cdf8cbc0cce41dba326a8ffb.sol:34:9: Warning: No visibility specified. Defaulting to "public".
        function requestClose(uint _balanceInitiator, uint _sequenceNumber, uint8 _v, bytes32 _r, bytes32 _s) {
        ^
Spanning multiple lines.
/tmp/tmp.aEq6oBGgwl/test_0a4d82f0af2c945e6a612af13a590776a772d071cdf8cbc0cce41dba326a8ffb.sol:46:9: Warning: No visibility specified. Defaulting to "public".
        function close()
        ^
Spanning multiple lines.
/tmp/tmp.aEq6oBGgwl/test_0a4d82f0af2c945e6a612af13a590776a772d071cdf8cbc0cce41dba326a8ffb.sol:15:13: Error: Undeclared identifier. Did you mean "initiator"?
            initator = msg.sender;
            ^------^
/tmp/tmp.aEq6oBGgwl/test_0a4d82f0af2c945e6a612af13a590776a772d071cdf8cbc0cce41dba326a8ffb.sol:49:70: Error: Undeclared identifier. Did you mean "counterpartyAddress"?
            require(now >= closeTimestamp + timeout || msg.sender == counteryPartyAddress(closeSequenceNumber + 1));
                                                                     ^------------------^
/tmp/tmp.aEq6oBGgwl/test_0a4d82f0af2c945e6a612af13a590776a772d071cdf8cbc0cce41dba326a8ffb.sol:52:17: Error: Undeclared identifier. Did you mean "initiator"?
            if (initator.send(closeBalanceInitiator))
                ^------^
```

File: docs/solidity-by-example.rst [hunk 7]
```diff
@@ -619,4 +619,63 @@ Safe Remote Purchase
...
+        uint closeBalanceInitiator;
+
+        enum State { Opened, Accepted, Closing, Closed }
+        State state;
+
+        function PaymentChannel(address _other, uint _timeout) payable {
+            initator = msg.sender;
+            acceptor = _other;
+            deposit = msg.value;
+            require(deposit < 2**250);
```
Comment (redsquirrel) [inline docs/solidity-by-example.rst:642]: I'm curious about this overflow check on msg.value. I wouldn't think to do that... is it really necessary?

Comment (chriseth) [reply docs/solidity-by-example.rst:642]: Not sure why I did this. Perhaps better safe than sorry?

**Comment by chriseth** [comment]
Some more comments are left to merge this. I actually would have liked to also include the frontend part. We can add a comment that this is still left to do.

**Comment by axic** [comment]
@bit-shift @ekpyron do you want to write some actual Solidity code ^^? 😉

**Comment by smarx** [comment]
@chriseth asked me to adapt my code from https://programtheblockchain.com/posts/2018/05/11/state-channels-for-two-player-games/ to go in the documentation.

I wonder, though, if the code from https://programtheblockchain.com/posts/2018/02/23/writing-a-simple-payment-channel/ (or perhaps https://programtheblockchain.com/posts/2018/03/02/building-long-lived-payment-channels/) might be better? That's a one-way payment channel, which is simpler to get right (and thus simpler code).

I guess I'm asking what the goal is here. If it's to get across the fundamentals of state channels, then maybe my state channel post is the most appropriate starting point. If it's to introduce payment channels, then I think my "simple payment channel" post is a good starting point.

If the goal is just to show how signatures can be used to make off-chain commitments, then https://programtheblockchain.com/posts/2018/02/17/signing-and-verifying-messages-in-ethereum/ is perhaps an even better starting point.

I'm happy to adapt any of those to a short example for the documentation, but I'd love some guidance on what people want to see.

**Comment by chriseth** [comment]
I think the goal is to show how easy it is to implement payment channels in etherem as opposed to bitcoin ;)

If a two-sided payment channel is too complicated, then a one-sided will do. It would be great if it could also include the required javascript code (or at least a snippet).

From a didactic point of view, it is probably a good idea to first have a short chapter on how to to create and verify signatures and then another chapter about how that can be used in payment channels.

**Comment by chriseth** [comment]
Added an issue, perhaps someone is willing to put up a bounty for this: #4131

**Comment by chriseth** [comment]
@smarx did you have time to work on this yet?

**Comment by smarx** [comment]
@chriseth No, sorry, I didn't. Looked like someone else is starting on the bounty, though!

**Comment by chriseth** [comment]
There are much better examples now.

**Comment by axic** [comment]
This was obsoleted by #4212.

**Pull request closed without merging**